### ISSUE TITLE
Segmentation Fault since String Length Increase

### ISSUE THREAD

**Issue by Sideboard** [libAtoms/QUIP#579]
`make test` fails on my laptop with the `gap_fit` tests. Turns out it is a segmentation fault introduced in #574. It seems to be caused by a stack overflow due to increasing the string lengths.

It works again if I use `ulimit -s unlimited` and the Github workflows also seem to be unbothered (looks like they have larger default stack sizes, https://github.com/actions/runner-images/issues/3257). But this may lead to confusing issues later. I think the job scripts of most people include `ulimit -s` unlimited but there is no information about that in the documentation.

@gabor1 I suppose the length increase is much larger than needed. Could we decrease it, what would work for you?

**Comment by gabor1** [comment]
first I increased them by a factor of 10, but that broke the CI for similar reasons, so I settled for a factor of 2. maybe the best route forward is to let this be controlled (or override the default) by some makefile directive ?

**Comment by Sideboard** [comment]
Even then, what should be the default?

And we better add some explanation to the documentation / install instructions. Under which circumstances should it be increased and to what?

Edit: And shall we keep all the different constants (`STRING_LENGTH`, `SYSTEM_STRING_LENGTH`, `*_LONG`)?

**Comment by gabor1** [comment]
I guess we could add the "ulimit -s unlimited" suggestion to the docs, and that would take care of most things. it’s not like its a huge memory issue that people are not actually going to have enough ram

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by Sideboard** [comment]
https://github.com/libAtoms/GAP/pull/82 gets rid of 6MB default stack usage.

We should still add the hint about `ulimit -s unlimited` to the docs.

**Comment by Sideboard** [comment]
With 30k each `%gap_str` is three times longer than `%command_line`, which is hard-coded to `len=10240`.

@gabor1 Are you using `config_file` for your input? The `config_string` is of type `extendable_str` and should not be limited.

**Comment by gabor1** [comment]
it's not me who's using it, it's all the other users I'm supporting via the issues and privately. Some use config_file some do not. what triggered these changes was not the command_line, but the dictionary value  (because gap="..." means the entire GAP string gets stored in a dictionary before it is parsed)

**Comment by gabor1** [comment]
I'm OK with leaving command_line hardcoded (but perhaps instead of a number it should be one of the *LENGTH constants), it would be nice to issue an error and suggesting config_file instead of crashing

**Comment by Sideboard** [comment]
I agree it should depend on one of the `*LENGTH` constants. What bothers me there is that even then anything that is parsed from `command_line` must be way shorter than `command_line` itself. (And are 200 gap entries really realisitc?)

I think we cannot detect a stack overflow per se as we cannot check how full the stack is. There is the function `increase_stacksize` in `system_module` but we would have to estimate what we need. I think this is already done for the Atoms.

We could also work with `extendable_str` or allocatable-length strings instead.

Edit: I think one reason for having `command_line` a specific length is that it's system-dependent how long the command line is allowed to be anyway.
```
$ getconf ARG_MAX        
2097152
```
